torchvision/datasets/_stereo_matching.py

```diff
@@ -0,0 +1,328 @@
+        self.transforms = transforms
+
+        self._images: List[Tuple[str, str]] = []
+        self._disparities: List[Tuple[str, str]] = []
+
+    def _read_img(self, file_path: str) -> Image.Image:
+        img = Image.open(file_path)
+        if img.mode != "RGB":
+            img = img.convert("RGB")
+        return img
+
+    def _scan_pairs(
+        self, paths_left_pattern: str, paths_right_pattern: str, fill_empty: bool = False
+    ) -> List[Tuple[str, str]]:
+        left_paths: List[str] = sorted(glob(paths_left_pattern))
+        right_paths: List[str] = sorted(glob(paths_right_pattern))
+
+        # used when dealing with inexistent disparity for the right image
+        if fill_empty:
+            right_paths = list("" for _ in left_paths)
+
+        if not left_paths:
+            raise FileNotFoundError(f"Could not find any files matching the patterns: {paths_left_pattern}")
+
+        if not right_paths:
+            raise FileNotFoundError(f"Could not find any files matching the patterns: {paths_right_pattern}")
+
+        if len(left_paths) != len(right_paths):
+            raise ValueError(
+                f"Found {len(left_paths)} left files but {len(right_paths)} right files using:\n "
+                f"left pattern: {paths_left_pattern}\n"
+                f"right pattern: {paths_right_pattern}\n"
+            )
+
+        images = list((left, right) for left, right in zip(left_paths, right_paths))
+        return images
+
+    @abstractmethod
+    def _read_disparity(self, file_path: str) -> Tuple:
+        # function that returns a disparity map and an occlusion map
+        pass
+
+    def __getitem__(self, index: int) -> Tuple:
+        """Return example at given index.
+        Args:
+            index(int): The index of the example to retrieve
+        Returns:
+            tuple: A 3 or 4-tuple with ``(img_left, img_right, disparity, Optional[valid_mask])`` where ``valid_mask``
+                can be a numpy boolean mask of shape (H, W) if the dataset provides a file
+                indicating which disparity pixels are valid. The disparity is a numpy array of
+                shape (1, H, W) and the images are PIL images. ``disparity`` is None for
+                datasets on which for ``split="test"`` the authors did not provide annotations.
+        """
+        img_left = self._read_img(self._images[index][0])
+        img_right = self._read_img(self._images[index][1])
+
+        dsp_map_left, valid_mask_left = self._read_disparity(self._disparities[index][0])
+        dsp_map_right, valid_mask_right = self._read_disparity(self._disparities[index][1])
+
+        imgs = (img_left, img_right)
+        dsp_maps = (dsp_map_left, dsp_map_right)
+        valid_masks = (valid_mask_left, valid_mask_right)
+
+        if self.transforms is not None:
+            (
+                imgs,
+                dsp_maps,
+                valid_masks,
+            ) = self.transforms(imgs, dsp_maps, valid_masks)
```

Comment:
I actually mostly wrapped up writing the transform pipeline yesterday, and I have mixed feelings about the signature now. Since tuples are immutable, sometimes you've got to do a lot of unpacking and repacking inside the transform, so I'd actually be more inclined to use the flattened signature, even though it comes at the cost of increased verbosity.

Reply:
@NicolasHug I'm missing context to provide an educated opinion. I understand that you are dealing with the legacy datasets and you have to "hack them together" to progress with the Depth Perception work, so in that sense I agree with your "keep as-is and re-evaluate" proposal.

I suspect that there are some incompatibilities with the current approach and the new Transforms API but that's not because you have multiple images/masks/maps but because left/right inputs need to be processed together rather than independently. I will probably need to understand more how this works on this task but it's an interesting use-case that we will need to support and might challenge some of the assumptions we made about the transforms and especially the fact the one that suggests that input can be transformed independently. cc @pmeier @vfdev-5

Reply:
To provide some context as to the specific use-case:

```python3
class RandomHorizontalFlip(torch.nn.Module):
    def __init__(self, p: float = 0.5) -> None:
        super().__init__()
        self.p = p

    def forward(
        self,
        images: Tuple[Tensor, Tensor],
        disparities: Tuple[Tensor, Tensor],
        masks: Tuple[Tensor, Tensor],
    ) -> Tuple[Tuple, Tuple, Tuple]:

        img_left, img_right = images
        dsp_left, dsp_right = disparities
        mask_left, mask_right = masks

        if dsp_right is not None and torch.rand(1) > self.p:
            # horizontal flip at pixel level
            img_left, img_right = F.hflip(img_left), F.hflip(img_right)
            dsp_left, dsp_right = F.hflip(dsp_left), F.hflip(dsp_right)
            # mask flipping flow is heavily skewed towards how we handle
            # generating pseudo-masks via the transform compose
            if mask_left is not None and mask_right is not None:
                mask_left, mask_right = F.hflip(mask_left), F.hflip(mask_right)
            # flip the two stereo channels
            return (
                (mask_right, mask_left),
                (dsp_right, dsp_left),
                (mask_right, mask_left)
            )

        return images, disparities, masks
```

The transform requires some inputs that for some datasets might not exist, namely the right disparity map. This is due to the fact, that we perform a horizontal pixel flip, the new target is not the transformed version of the original target but the opposite channel flipped target. Therefor the transform needs to have access to all the information.

Were it not for this transform, we could have just abided by the previous flow template.

Reply:
For the bigger picture view, here is some extra context:

```python3
class Compose(torch.nn.Module):
    def __init__(self, transforms):
        super().__init__()
        self.transforms = transforms

    def forward(self, images, disparities, masks):
        for t in self.transforms:
            images, disparities, masks = t(images, disparities, masks)
        return images, disparities, masks
```

Flattening everything would require making transforms that support the following `forward` scheme:
```python3
    def forward(self, img_left, img_right, dsp_left, dsp_right, mask_left, mask_right):
        for t in self.transforms:
            img_left, img_right, dsp_left, dsp_right, mask_left, mask_right = t(img_left, img_right, dsp_left, dsp_right, mask_left, mask_right)
        return img_left, img_right, dsp_left, dsp_right, mask_left, mask_right
```

Reply:
I am not very familiarised with the new transforms API, but looking into [it](https://github.com/pytorch/vision/blob/main/torchvision/prototype/transforms/_transform.py#L11) could we pass everything as a dictionary and return also a dictionary?

Reply:
I'm not sure how much this is considered canonical workflow. Dicts however were my personal go-to choice when faced with not-so straight-forward data workflows.

```python3
stereo_dataset_inputs = {
    "left_image": ...,
    "right_image": ...,
    "left_disparity": ...,
    "right_disparity": ...,
    "left_mask": ...,
    "right_mask": ...,
}
stereo_dataset_inputs = transform(stereo_dataset_inputs)
return stereo_dataset_inputs["left_image"], ...
```

HuggingFace does something similar for [example](https://huggingface.co/datasets/wikitext)

Reply:
Yes that kind of transforms are supported. But keep in mind that you will need to create custom classes that implement the logic the way you describe above. The API gives you goodies/utilities to assist your development.

Another thing you will need to do is possibly define a new custom type for Disparity Maps. We already offer support for images and masks and thankfully it's easy to define new types. There you would need to define what each operator can do to the maps and how they work (example: can you resize it using bilinear? how are colour shifts on images affect disparity maps? etc).

Reply:
Truthfully, a **disparity map is actually just a flow map** with a single channel / **0-flow on the y-axis**. The same constraints that apply to the transforms in our optical flow reference scripts apply to these as well.

Bilinear resizing will still work as intended because both flows and disparities are measured in pixels and not some other real-world distance metric.

Judging by what several other paper repos provide in terms of augmentation pipelines ([1](https://github.com/megvii-research/CREStereo/blob/master/dataset.py) , [2](https://github.com/princeton-vl/RAFT-Stereo/blob/main/core/stereo_datasets.py)) it seems that the same approach is being followed.

Typically in monocular depth-estimation the same problems are posed and although [lambertian surfaces](https://en.wikipedia.org/wiki/Lambertian_reflectance) are an important issue, the assumption that color shifts should affect the depth-map are not taken into account. Both in the unsupervised and supervised setting we consider such alterations something that should not affect the **true world** model / estimation, but they are solely camera related / a regularization method.

Reply:
In any case I agree that we can merge this PR to unblock and re-visit this considering the new transforms API.

Reply:
OK sounds like the new API in principle supports both the multiple inputs (named or not) and special types like disparity maps. I don't see immediately an incompatibility but we should keep an eye for similar requests as this is a useful feedback for the new API.

Reply:
@datumbox @NicolasHug since, since this PR is non-blocking for the model training, I could try dog-fooding the new transforms API and include in this PR parts of the reference script in order for us to get a clearer idea how everything would mix in with the new API and have a better picture of how the dataset should be structured. 

Is that ok with you guys?

Reply:
We literally just merged a bunch of updates today at #6305. It might be a bit premature to dogfood it (we'll definitely take upon your offer a bit later though) but you can already have a look how it works on main branch at `torchvision.prototype.transforms`.